src/webgpu/api/operation/command_buffer/CopyTextureToTexture.spec.ts

```diff
@@ -0,0 +1,301 @@
+      textureSize.srcTextureSize,
+      format,
+      srcCopyLevel
+    );
+    const srcTextureSizeAtLevel = physicalMipSize(
+      textureSize.srcTextureSize,
+      format,
+      '2d',
+      srcCopyLevel
+    );
+    const bytesPerBlock = kAllTextureFormatInfo[format].bytesPerBlock;
+    assert(bytesPerBlock !== undefined);
+    t.device.defaultQueue.writeTexture(
+      { texture: srcTexture, mipLevel: srcCopyLevel },
+      initialSrcData,
+      {
+        offset: 0,
+        bytesPerRow: srcTextureSizeAtLevel.width * bytesPerBlock,
+        rowsPerImage: srcTextureSizeAtLevel.height,
+      },
+      srcTextureSizeAtLevel
+    );
+
+    // Copy the region specified by copyBoxOffsets from srcTexture to dstTexture.
+    const dstTextureSizeAtLevel = physicalMipSize(
+      textureSize.dstTextureSize,
+      format,
+      '2d',
+      dstCopyLevel
+    );
+    const minWidth = Math.min(srcTextureSizeAtLevel.width, dstTextureSizeAtLevel.width);
+    const appliedCopyWidth =
+      minWidth +
+      copyBoxOffsets.copyExtent.width -
+      Math.max(copyBoxOffsets.srcOffset.x, copyBoxOffsets.dstOffset.x);
+    const minHeight = Math.min(srcTextureSizeAtLevel.height, dstTextureSizeAtLevel.height);
+    const appliedCopyHeight =
+      minHeight +
+      copyBoxOffsets.copyExtent.height -
+      Math.max(copyBoxOffsets.srcOffset.y, copyBoxOffsets.dstOffset.y);
+    const appliedCopyDepth = Math.max(
+      textureSize.srcTextureSize.depth +
+        copyBoxOffsets.copyExtent.depth -
+        Math.max(copyBoxOffsets.srcOffset.z, copyBoxOffsets.dstOffset.z),
+      0
+    );
+
+    const encoder = t.device.createCommandEncoder();
+    encoder.copyTextureToTexture(
+      { texture: srcTexture, mipLevel: srcCopyLevel, origin: copyBoxOffsets.srcOffset },
+      { texture: dstTexture, mipLevel: dstCopyLevel, origin: copyBoxOffsets.dstOffset },
+      { width: appliedCopyWidth, height: appliedCopyHeight, depth: appliedCopyDepth }
+    );
+
+    // Copy the whole content of dstTexture at dstCopyLevel to dstBuffer.
+    const bytesPerDstAlignedRow = align(dstTextureSizeAtLevel.width * bytesPerBlock, 256);
+    const dstBufferSize =
+      (dstTextureSizeAtLevel.width * dstTextureSizeAtLevel.height - 1) * bytesPerDstAlignedRow +
+      align(dstTextureSizeAtLevel.width * bytesPerBlock, 4);
+    const dstBufferDesc: GPUBufferDescriptor = {
+      size: dstBufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    };
+    const dstBuffer = t.device.createBuffer(dstBufferDesc);
+
+    encoder.copyTextureToBuffer(
+      { texture: dstTexture, mipLevel: dstCopyLevel, origin: { x: 0, y: 0, z: 0 } },
+      {
+        buffer: dstBuffer,
+        bytesPerRow: bytesPerDstAlignedRow,
+        rowsPerImage: dstTextureSizeAtLevel.height,
+      },
+      dstTextureSizeAtLevel
+    );
+    t.device.defaultQueue.submit([encoder.finish()]);
+
+    // Fill expectedDataWithPadding with the expected data of dstTexture.
```

Comment:
nit: comment that this checks the texels untouched by the copy are zero (their previous value.)